lib/si-data-nats/src/jetstream/context.rs

```diff
@@ -0,0 +1,136 @@
+        let name = name.into();
+        validate_stream_subject_name(&name)?;
+
+        let stream = self
+            .inner
+            .get_or_create_stream(Config {
+                name,
+                subjects,
+                max_messages: DEFAULT_MAX_MESSAGES,
+                // TODO(nick): for the rebaser, we must have a work queue retention policy. This is temporary, just for
+                // the first pass. I hope this comment doesn't come back to haunt me.
+                // retention: RetentionPolicy::WorkQueue,
+                ..Default::default()
+            })
+            .await?;
+        Ok(stream)
+    }
+
+    /// Publishes to a stream with a reply mailbox in the headers. Immediately awaits the future that the server has
+    /// acknowledged the message.
+    pub async fn publish_with_reply_mailbox_and_immediately_ack(
+        &self,
+        client: &Client,
+        subject: String,
+        bytes: Bytes,
+    ) -> JetstreamResult<String> {
+        let mut headers = HeaderMap::new();
+        let reply_subject = client.new_inbox();
+        headers.insert(REPLY_SUBJECT_HEADER_NAME.clone(), reply_subject.as_str());
+
+        let ack_future = self
+            .inner
+            .publish_with_headers(subject, headers, bytes)
+            .await?;
+
+        ack_future.await?;
+
+        Ok(reply_subject)
+    }
+
+    /// Finds or creates a [`Consumer`] with a durable, pull configuration. Because the consumer is durable, the server
+    /// will remember the last events "acked" in the event of a failure.
+    pub async fn get_or_create_durable_consumer(
+        &self,
+        stream: &Stream,
+        name: impl Into<String>,
+    ) -> JetstreamResult<Consumer> {
+        let name = name.into();
+        let raw_consumer = stream
+            .get_or_create_consumer(
+                name.as_str(),
+                jetstream::consumer::pull::Config {
+                    durable_name: Some(name.clone()),
+                    ..Default::default()
+                },
+            )
+            .await?;
+        Ok(Consumer::new(raw_consumer))
+    }
+}
+
+// This is an opinionated version of the rules from upstream. For example, alphanumeric characters are "recommended",
+// but we use "-" for subject prefixes for testing since we cannot use "." (it is prohibited and you will fail stream
+// creation).
+//
+// Link to upstream docs: https://docs.nats.io/running-a-nats-service/nats_admin/jetstream_admin/naming
+fn validate_stream_subject_name(subject: impl AsRef<str>) -> JetstreamResult<()> {
+    let subject = subject.as_ref();
+    for char in subject.chars() {
+        let valid = match char {
+            ' ' | '.' | '>' | '*' | '/' | '\\' => false,
+            char if char != '-' && !char.is_alphanumeric() => false,
+            _ => true,
+        };
+        if !valid {
+            return Err(JetstreamError::InvalidSubjectName(subject.to_string()));
+        }
+    }
+    Ok(())
+}
```

Comment:
Without this, we would run into infinite hangs for tests. Perhaps, that's an upstream issue with the client. The prefix cannot be joined by a '.' character (causes the hang), so we use a '-' character.

However, stream name sanitation isn't the worst idea either. We should likely have a longer-term vision for how we want to centralize and construct NATS subjects across SI.

Reply:
Update: filed an issue https://github.com/nats-io/nats.rs/issues/1213